src/meta/src/hummock/manager/tests.rs

```diff
@@ -939,6 +940,124 @@ async fn test_trigger_manual_compaction() {
+    env.hash_mapping_manager()
+        .build_fragment_hash_mapping(1, &parallel_units);
+    for table_id in 1..sst_num + 2 {
+        env.hash_mapping_manager()
+            .set_fragment_state_table(1, table_id as u32);
+    }
+
+    // No compaction task available.
+    let task = hummock_manager
+        .get_compact_task(StaticCompactionGroupId::StateDefault.into())
+        .await
+        .unwrap();
+    assert_eq!(task, None);
+
+    // Add some sstables and commit.
+    let epoch: u64 = 1;
+    let original_tables = generate_test_tables(epoch, get_sst_ids(&hummock_manager, sst_num).await);
+    register_sstable_infos_to_compaction_group(
+        hummock_manager.compaction_group_manager_ref_for_test(),
+        &original_tables,
+        StaticCompactionGroupId::StateDefault.into(),
+    )
+    .await;
+    commit_from_meta_node(
+        hummock_manager.borrow(),
+        epoch,
+        to_local_sstable_info(&original_tables),
+    )
+    .await
+    .unwrap();
+
+    // Get a compaction task.
+    let mut compact_task = hummock_manager
+        .get_compact_task(StaticCompactionGroupId::StateDefault.into())
+        .await
+        .unwrap()
+        .unwrap();
+    hummock_manager
+        .assign_compaction_task(&compact_task, context_id, async { true })
+        .await
+        .unwrap();
+    assert_eq!(
+        compact_task
+            .get_input_ssts()
+            .first()
+            .unwrap()
+            .get_level_idx(),
+        0
+    );
+    assert_eq!(compact_task.get_task_id(), 2);
+    // send task
+    compactor_manager
+        .random_compactor()
+        .unwrap()
+        .send_task(Task::CompactTask(compact_task.clone()))
+        .await
+        .unwrap();
+
+    // send heartbeats to the task immediately
+    let req = CompactTaskProgress {
+        task_id: compact_task.task_id,
+        num_blocks_sealed: 1,
+        num_blocks_uploaded: 1,
+    };
+    compactor_manager.update_compaction_task_progress(context_id, vec![req]);
+
+    // Cancel the task immediately and succeed.
+    compact_task.task_status = false;
+    assert!(hummock_manager
+        .report_compact_task(context_id, &compact_task)
+        .await
+        .unwrap());
+
+    // Get a compaction task.
+    let mut compact_task = hummock_manager
+        .get_compact_task(StaticCompactionGroupId::StateDefault.into())
+        .await
+        .unwrap()
+        .unwrap();
+    hummock_manager
+        .assign_compaction_task(&compact_task, context_id, async { true })
+        .await
+        .unwrap();
+    assert_eq!(compact_task.get_task_id(), 3);
+    // send task
+    compactor_manager
+        .random_compactor()
+        .unwrap()
+        .send_task(Task::CompactTask(compact_task.clone()))
+        .await
+        .unwrap();
+
+    // do not send heartbeats to the task for 2.5 seconds (ttl = 1s, heartbeat check freq. = 1s)
+    tokio::time::sleep(std::time::Duration::from_millis(2500)).await;
```

Comment:
Time dependent tests can be flaky. But I think it won't be a problem in [deterministic test](https://github.com/singularity-data/risingwave/pull/4476).
Maybe we want an attribute macro that only enables test_hummock_compaction_task_heartbeat in deterministic test?

Reply:
Maybe `#[cfg(madsim)]` @wangrunji0408 ?

Reply:
Yes, just add it to this test.